compiler/damlc/daml-stdlib-src/DA/Ord/Builtin.daml

```diff
@@ -0,0 +1,125 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE CPP #-}
+
+#ifndef DAML_GENMAP
+module DA.Ord.Builtin where
+#else
+
+-- | Note: This is only supported in DAML-LF 1.11 or later.
+--
+-- This module provides variants of other standard library
+-- functions that are based on the builtin Daml-LF ordering rather
+-- than user-defined ordering. This is the same order also used
+-- by `DA.Map`.
+--
+-- These functions are usually much more efficient than their
+-- `Ord`-based counterparts.
+--
+-- Note that the functions in this module still require `Ord`
+-- constraints. This is purely to enforce that you don’t
+-- pass in values that cannot be compared, e.g., functions. The
+-- implementation of those instances is not used.
+module DA.Ord.Builtin
+  ( dedup
+  , dedupOn
+  , dedupSort
+  , dedupOnSort
+  , sort
+  , sortOn
+  ) where
+
+-- import DA.Map (Map)
+import qualified DA.Map as Map
+
+-- | `dedup l` removes duplicate elements from a list. In particular,
+-- it keeps only the first occurence of each element.
+--
+-- `dedup` is stable so the elements in the output are ordered
+-- by their first occurence in the input. If you do not need
+-- stability, consider using `dedupSort` which is more efficient.
+--
+-- ```
+-- >>> dedup [3, 1, 1, 3]
+-- [3, 1]
+-- ```
+dedup : Ord a => [a] -> [a]
+dedup = dedupOn identity
+
+-- | A version of `dedup` where deduplication is done
+-- after applying the given function. Example use: `dedupOn (.employeeNo) employees`.
+--
+-- `dedupOn` is stable so the elements in the output are ordered
+-- by their first occurence in the input. If you do not need
+-- stability, consider using `dedupOnSort` which is more efficient.
+--
+-- ```
+-- >>> dedupOn fst [(3, "a"), (1, "b"), (1, "c"), (3, "d")]
+-- [(3, "a"), (1, "b")]
+-- ```
+dedupOn f xs = Map.values (Map.fromList (Map.values deduped))
+  where
+    -- Fused Map.fromListWith + map/fold
+    deduped = snd (foldl insert (0, Map.empty) xs)
+    insert (i, m) x =
+      let k = f x
+          m' = case Map.lookup k m of
+            Some prev@(j, _)
+              | j <= i -> m
+            _ -> Map.insert k (i, x) m
+      in (i + 1, m')
+
+-- | `dedupSort` is a more efficient variant of `dedup`
+-- that does not preserve the order of the input elements.
+-- Instead the output will be sorted acoording to the builtin Daml-LF
+-- ordering.
+--
+-- ```
+-- >>> dedupSort [3, 1, 1, 3]
+-- [1, 3]
+-- ```
+dedupSort : Ord a => [a] -> [a]
+dedupSort = dedupOnSort identity
+
+-- | `dedupOnSort` is a more efficient variant of `dedupOn`
+-- that does not preserve the order of the input elements.
+-- Instead the output will be sorted on the values returned by the function.
+--
+-- For duplicates, the last element in the list will be included in the output.
```

Comment:
I’m slightly unhappy about this. It seems nicer if `dedupOnSort` wouldn’t also change whether we preserve the first or the last element. We can easily change that but it means we pay for an extra lookup before the insert. I don’t have particularly strong feelings here either way and I don’t have the microbenchmarks to classify the impact so happy to change it either way.

Reply:
You could use a foldr to make it preserve the first element instead of the last. (I imagine foldr vs foldl is about the same in terms of impact, compared to doing a lookup.)

Reply:
duh, fixed in 1214abe31ce3c28fbcd4d94054ed66aa2a6ae7bd